File: psisearch2/clustal2fasta.pl
#!/usr/bin/env perl

################################################################
# copyright (c) 2014,2015 by William R. Pearson and The Rector &
# Visitors of the University of Virginia */
################################################################
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under this License is distributed on an "AS
# IS" BASIS, WITHOUT WRRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied.  See the License for the specific language
# governing permissions and limitations under the License.
################################################################

################################################################
# clustal2fasta.pl 
################################################################
# clustal2fasta.pl takes a standard clustal format alignment file
# and produces the corresponding FASTA file.
#
################################################################

use warnings;
use strict;
use Pod::Usage;
use Getopt::Long;

my ($shelp, $help, $trim) = (0, 0);

GetOptions(
    "h|?" => \$shelp,
    "help" => \$help,
    );

pod2usage(1) if $shelp;
pod2usage(exitstatus => 0, verbose => 2) if $help;
unless (-f STDIN || -p STDIN || @ARGV) {
 pod2usage(1);
}

my @seq_ids = ();
my %msa = ();
    
# read the first line, first should not be blank
my $title = <>;

while (my $line = <>) {
  chomp $line;
  next unless ($line);
  next if ($line =~ m/^[\s:\*\+\.]+$/);   # skip conservation line

  my ($seq_id, $align) = split(/\s+/,$line);

  if (defined($msa{$seq_id})) {
    $msa{$seq_id} .= $align;
  }
  else {
    $msa{$seq_id} = $align;
    push @seq_ids, $seq_id;
  }
}

for my $seq_id ( @seq_ids ) {
  my $fmt_seq = $msa{$seq_id};
  $fmt_seq =~ s/(.{0,60})/$1\n/g;
  print ">$seq_id\n$fmt_seq";
}

__END__

=pod

=head1 NAME

 clustal2fasta.pl

=head1 SYNOPSIS

 clustal2fasta.pl clustal.msa

=head1 OPTIONS

 -h	short help
 --help include description


=head1 DESCRIPTION

C<clustal2fasta.pl> takes a Clustal format interleaved multiple
sequence alignment and produces the corresponding fasta format library.

=head1 AUTHOR

William R. Pearson, wrp@virginia.edu

=cut

File: psisearch2/clustal2fasta.py
#!/usr/bin/env python

################################################################
# copyright (c) 2014,2015 by William R. Pearson and The Rector &
# Visitors of the University of Virginia */
################################################################
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under this License is distributed on an "AS
# IS" BASIS, WITHOUT WRRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied.  See the License for the specific language
# governing permissions and limitations under the License.
################################################################

################################################################
# clustal2fasta.pl 
################################################################
# clustal2fasta.pl takes a standard clustal format alignment file
# and produces the corresponding FASTA file.
#
# if --end_mask or --int_mask are set, then end or internal '-'s are converted to the query (first) sequence
# if --trim is set, then alignments beyond the beginning/end of the query sequence are trimmed
#
################################################################

import argparse
import fileinput
import re

################
#
# python re-write of clustal2fasta.pl
#
# in the future, modify for various query seeding strategies
################    

arg_parse = argparse.ArgumentParser(description='Convert clustal MSA to FASTA library')
arg_parse.add_argument('--query|--query_file', dest='query_file', action='store',help='query sequence file')
arg_parse.add_argument('files', metavar='FILE', nargs='*', help='files to read, if empty, stdin is used')
args=arg_parse.parse_args()

msa = {}
seq_ids = []

is_line1 = True
for line in fileinput.input(args.files):
    if is_line1:
        is_line1 = False
        continue
    line = line.strip()
    if not line:
        continue
    if re.search(r'^[\s:\*\+\.]+$',line):
        continue
    
    (seq_id, align) = re.split(r'\s+',line)

    if seq_id in msa:
        msa[seq_id] += align
    else:
        msa[seq_id] = align
        seq_ids.append(seq_id)

for seq_id in seq_ids:
    fmt_seq = re.sub(r'(.{0,60})',r'\1\n',msa[seq_id])
    print ">%s\n%s" % (seq_id, fmt_seq)
